File: data/langpack/patch_langpack.go
package main

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"regexp"
	"strings"
)

type LanguagePack struct {
	Schema int `json:"schema"`
	Packs  []struct {
		Name      string `json:"name"`
		Languages []struct {
			LangCode        string `json:"lang_code"`
			Name            string `json:"name"`
			NativeName      string `json:"native_name"`
			PluralCode      string `json:"plural_code"`
			Official        bool   `json:"official"`
			Rtl             bool   `json:"rtl,omitempty"`
			StringsCount    int    `json:"strings_count"`
			TranslatedCount int    `json:"translated_count"`
			TranslationsUrl string `json:"translations_url"`
			Version         int    `json:"version"`
			File            string `json:"file"`
			Sha256          string `json:"sha256"`
		} `json:"languages"`
	} `json:"packs"`
}

func fileSHA256(path string) (string, error) {
	f, err := os.Open(path)
	if err != nil {
		return "", err
	}
	defer f.Close()

	h := sha256.New()
	if _, err := io.Copy(h, f); err != nil {
		return "", err
	}

	return hex.EncodeToString(h.Sum(nil)), nil
}

func main() {
	baseDir := `c:\Users\myahm\OneDrive\Documents\Whatsgram\gramsrv\data\langpack`
	jsonPath := filepath.Join(baseDir, "official-language-packs.json")

	b, err := os.ReadFile(jsonPath)
	if err != nil {
		panic(err)
	}

	var data LanguagePack
	if err := json.Unmarshal(b, &data); err != nil {
		panic(err)
	}

	versionRe := regexp.MustCompile(`_v(\d+)\.strings$`)
	updatedCount := 0

	for i := range data.Packs {
		pack := &data.Packs[i]
		for j := range pack.Languages {
			lang := &pack.Languages[j]
			if lang.File == "" {
				continue
			}

			// Only process Android for speed, or remove this to process all
			if pack.Name != "android" && pack.Name != "android_x" {
				continue
			}

			absFile := filepath.Join(baseDir, lang.File)
			contentBytes, err := os.ReadFile(absFile)
			if err != nil {
				fmt.Printf("File not found: %s\n", absFile)
				continue
			}
			content := string(contentBytes)

			newContent := strings.ReplaceAll(content, "Telegram", "Whatsgram")
			newContent = strings.ReplaceAll(newContent, "telegram", "whatsgram")
			newContent = strings.ReplaceAll(newContent, "TELEGRAM", "WHATSGRAM")

			if newContent == content {
				continue
			}

			lang.Version++
			newRelFile := versionRe.ReplaceAllString(lang.File, fmt.Sprintf("_v%d.strings", lang.Version))
			if newRelFile == lang.File {
				newRelFile = strings.Replace(lang.File, ".strings", fmt.Sprintf("_v%d.strings", lang.Version), 1)
			}
			newAbsFile := filepath.Join(baseDir, newRelFile)

			if err := os.WriteFile(newAbsFile, []byte(newContent), 0644); err != nil {
				panic(err)
			}
			if newAbsFile != absFile {
				os.Remove(absFile)
			}

			hash, err := fileSHA256(newAbsFile)
			if err != nil {
				panic(err)
			}

			lang.File = newRelFile
			lang.Sha256 = hash

			fmt.Printf("Updated %s -> %s\n", lang.Name, newRelFile)
			updatedCount++
		}
	}

	if updatedCount > 0 {
		outBytes, err := json.MarshalIndent(data, "", "  ")
		if err != nil {
			panic(err)
		}
		if err := os.WriteFile(jsonPath, outBytes, 0644); err != nil {
			panic(err)
		}
		fmt.Printf("Successfully updated %d language packs and official-language-packs.json!\n", updatedCount)
	} else {
		fmt.Println("No changes needed.")
	}
}

File: data/langpack/patch_langpack.py
import json
import os
import hashlib
import re

def sha256_file(filepath):
    sha256 = hashlib.sha256()
    with open(filepath, 'rb') as f:
        for chunk in iter(lambda: f.read(4096), b""):
            sha256.update(chunk)
    return sha256.hexdigest()

def patch_langpacks(base_dir):
    json_path = os.path.join(base_dir, 'official-language-packs.json')
    with open(json_path, 'r', encoding='utf-8') as f:
        data = json.load(f)

    for pack in data.get('packs', []):
        for lang in pack.get('languages', []):
            rel_file = lang.get('file')
            if not rel_file:
                continue
            
            abs_file = os.path.join(base_dir, rel_file)
            if not os.path.exists(abs_file):
                print(f"File not found: {abs_file}")
                continue
                
            # Read and replace content
            with open(abs_file, 'r', encoding='utf-8') as f:
                content = f.read()
                
            # Replace Telegram with Whatsgram (case sensitive)
            new_content = content.replace("Telegram", "Whatsgram")
            new_content = new_content.replace("telegram", "whatsgram")
            new_content = new_content.replace("TELEGRAM", "WHATSGRAM")
            
            # If no changes, skip to avoid unnecessary bumps
            if new_content == content:
                continue
                
            # Increment version
            old_version = lang.get('version', 0)
            new_version = old_version + 1
            lang['version'] = new_version
            
            # Form new filename
            # e.g., android/android_en_v59885528.strings -> android/android_en_v59885529.strings
            new_rel_file = re.sub(r'_v\d+\.strings$', f'_v{new_version}.strings', rel_file)
            
            # If regex didn't match, just append
            if new_rel_file == rel_file:
                new_rel_file = rel_file.replace('.strings', f'_v{new_version}.strings')
                
            new_abs_file = os.path.join(base_dir, new_rel_file)
            
            # Write new file
            with open(new_abs_file, 'w', encoding='utf-8') as f:
                f.write(new_content)
                
            # Remove old file if name changed
            if new_abs_file != abs_file:
                os.remove(abs_file)
                
            # Update json entry
            lang['file'] = new_rel_file
            lang['sha256'] = sha256_file(new_abs_file)
            
            print(f"Updated {rel_file} -> {new_rel_file}")

    # Write back json
    with open(json_path, 'w', encoding='utf-8') as f:
        json.dump(data, f, indent=2, ensure_ascii=False)
        
    print("Done patching langpacks!")

if __name__ == '__main__':
    base_dir = r"c:\Users\myahm\OneDrive\Documents\Whatsgram\gramsrv\data\langpack"
    patch_langpacks(base_dir)
